lib/Internal/EmitSource.php

```diff
@@ -0,0 +1,302 @@
+<?php
+
+namespace Amp\Internal;
+
+use Amp\Deferred;
+use Amp\DisposedException;
+use Amp\Failure;
+use Amp\Promise;
+use Amp\Stream;
+use Amp\Success;
+use React\Promise\PromiseInterface as ReactPromise;
+
+/**
+ * Class used internally by {@see Stream} implementations. Do not use this class in your code, instead compose your
+ * class from one of the available classes implementing {@see Stream}.
+ *
+ * @internal
+ *
+ * @template TValue
+ * @template TSend
+ */
+final class EmitSource
+{
+    /** @var Promise|null */
+    private $result;
+
+    /** @var bool */
+    private $completed = false;
+
+    /** @var mixed[] */
+    private $emittedValues = [];
+
+    /** @var Promise[] */
+    private $sendValues = [];
+
+    /** @var Deferred[] */
+    private $backPressure = [];
+
+    /** @var Deferred[] */
+    private $waiting = [];
+
+    /** @var int */
+    private $consumePosition = 0;
+
+    /** @var int */
+    private $emitPosition = 0;
+
+    /** @var array|null */
+    private $resolutionTrace;
+
+    /** @var bool */
+    private $disposed = false;
+
+    /** @var bool */
+    private $used = false;
+
+    /**
+     * @return Promise<mixed|null>
+     *
+     * @psalm-return Promise<TValue|null>
+     */
+    public function continue(): Promise
+    {
+        return $this->next(new Success);
```

Comment:
We're creating a lot of unnecessary objects here, maybe store the `new Success` in a static field, as it's basically immutable?

Reply:
Yeah, I could definitely optimize that away.